x-pack/plugins/security_solution/server/endpoint/services/metadata/endpoint_metadata_service.ts

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import {
+  ElasticsearchClient,
+  KibanaRequest,
+  Logger,
+  SavedObjectsClientContract,
+  SavedObjectsServiceStart,
+} from 'kibana/server';
+import { HostInfo, HostMetadata } from '../../../../common/endpoint/types';
+import { Agent, AgentPolicy, PackagePolicy } from '../../../../../fleet/common';
+import {
+  AgentNotFoundError,
+  AgentPolicyServiceInterface,
+  AgentService,
+} from '../../../../../fleet/server';
+import {
+  EndpointHostNotFoundError,
+  EndpointHostUnEnrolledError,
+  FleetAgentNotFoundError,
+  FleetAgentPolicyNotFoundError,
+} from './errors';
+import { getESQueryHostMetadataByID } from '../../routes/metadata/query_builders';
+import { queryResponseToHostResult } from '../../routes/metadata/support/query_strategies';
+import { DEFAULT_ENDPOINT_HOST_STATUS, fleetAgentStatusToEndpointHostStatus } from '../../utils';
+import { EndpointError } from '../../errors';
+
+type AgentPolicyWithPackagePolicies = Omit<AgentPolicy, 'package_policies'> & {
+  package_policies: PackagePolicy[];
+};
+
+// Will wrap the given Error with `EndpointError`, which will
+// help getting a good picture of where in our code the error originated from.
+const wrapErrorIfNeeded = (error: Error): EndpointError =>
+  error instanceof EndpointError ? error : new EndpointError(error.message, error);
+
+// used as the callback to `Promise#catch()` to ensure errors
+// (especially those from kibana/elasticsearch clients) are wrapped
+const catchAndWrapError = <E extends Error>(error: E) => Promise.reject(wrapErrorIfNeeded(error));
+
+export class EndpointMetadataService {
+  /**
+   * For internal use only by the `this.DANGEROUS_INTERNAL_SO_CLIENT`
+   * @deprecated
+   */
+  private __DANGEROUS_INTERNAL_SO_CLIENT: SavedObjectsClientContract | undefined;
+
+  constructor(
+    private savedObjectsStart: SavedObjectsServiceStart,
+    private readonly agentService: AgentService,
+    private readonly agentPolicyService: AgentPolicyServiceInterface,
+    private readonly logger?: Logger
+  ) {}
+
+  /**
+   * A Saved Object client that can access any saved object. Used primarly to retrieve fleet data
+   * for endpoint enrichment (so that users are not required to have superuser role)
```

Comment:
1: thanks for the scary property name

2: can you add to this comment _why_ it is dangerous? A note that it is effectively the **system user** and has all privileges and permissions.   "can access any object" sounds like it can just _read_ extra stuff you might not want. But this can modify and delete anything as well.

Reply:
Sure, I can add more to it.

Or....

Question: should I make this allot more restrictive in that only the methods we are using will be exposed _(I think I might able to do that with a Proxy)_? meaning: if someone try to use a `.create()` from this internal SO client, it would `throw` an error.   Thoughts?

Right now, we're only using the .`get()` method.

@pzl  ☝️